### ISSUE TITLE
Python pipeline support

### ISSUE THREAD

**Issue by brollb** [deepforge-dev/deepforge#1024]
Migrate from lua and torch7 to python (and keras/tensorflow/pytorch) for v2.x

**Comment by brollb** [comment]
It may be better to consider an architecture enabling support for TF as well (as a separately module)...

**Comment by brollb** [comment]
This pull request is pretty major and includes a pretty noteworthy change to the architecture of DeepForge. Currently, the base projects include two main libraries (behind the scenes): _pipeline_ and _nn_. Where the _nn_ library is the Torch7 neural network package.

This pull request converts the pipeline and operation support to use python as their implementation language and no longer includes any neural network library in the basic project seed. PyTorch/TF can then be defined as separate _modules_ (containing a data model, visual interface and code generators) which can be then loaded and used by the given user. This also enables other modules to be created for other things like mxnet or even scikit learn concepts.

This PR also addresses another issue with the type system. One minor annoyance (to me anyway) in version 1.x was the lack of a strong binding between the types defined in DeepForge (for use in the pipelines) and the types in the underlying language. That is, the types defined in the pipelines were not necessary bound to types in torch7. This is being changed for v2.0.0 so that the types are simply annotations that are parsed and provided as hints in the pipeline editor.

**Comment by brollb** [comment]
As stated in the previous comment, the neural network support can be provided as a separate extension to deepforge allowing support for multiple different python neural network libraries rather than requiring one specific one.

This will mean that this repo can be restructured to no longer contain any of the neural network support and that support can be developed separately (in https://github.com/deepforge-dev/deepforge-keras). As this is the case, I will be changing this issue to reflect this and only focus on the development of the pipeline system and not pytorch support (which can be developed in another repo as is being done with the earlier example.

**Comment by brollb** [comment]
Added to v2.0.0 (still very much in alpha)
